### PR TITLE
[ATen][NATIVE][CUDA] Allow all 10.x compute capabilities for using vec8 kernel

### PR DESCRIPTION
**Pull request by Aidyn-A**
This will allow `sm_103` devices call vec8 kernels.
Verification script:
```Python
import torch
from torch.profiler import profile, ProfilerActivity

device = torch.device("cuda")

for dtype in (torch.bfloat16, torch.float16,):
    x = torch.randn(1024, device=device, dtype=dtype)
    with profile(activities=[ProfilerActivity.CUDA], record_shapes=True) as prof:
        y = torch.relu(x)
    stats = prof.key_averages()
    for entry in stats:
        if "at::native::vectorized_elementwise_kernel" in entry.key:
            print(entry.key)
```

Before:
```
void at::native::vectorized_elementwise_kernel<4, at::native::(anonymous namespace)::launch_clamp_scalar(at::TensorIteratorBase&, c10::Scalar, c10::Scalar, at::native::detail::ClampLimits)::{lambda()#1}::operator()() const::{lambda()#9}::operator()() const::{lambda(c10::BFloat16)#1}, std::array<char*, 2ul> >(int, at::native::(anonymous namespace)::launch_clamp_scalar(at::TensorIteratorBase&, c10::Scalar, c10::Scalar, at::native::detail::ClampLimits)::{lambda()#1}::operator()() const::{lambda()#9}::operator()() const::{lambda(c10::BFloat16)#1}, std::array<char*, 2ul>)
void at::native::vectorized_elementwise_kernel<4, at::native::(anonymous namespace)::launch_clamp_scalar(at::TensorIteratorBase&, c10::Scalar, c10::Scalar, at::native::detail::ClampLimits)::{lambda()#1}::operator()() const::{lambda()#8}::operator()() const::{lambda(c10::Half)#1}, std::array<char*, 2ul> >(int, at::native::(anonymous namespace)::launch_clamp_scalar(at::TensorIteratorBase&, c10::Scalar, c10::Scalar, at::native::detail::ClampLimits)::{lambda()#1}::operator()() const::{lambda()#8}::operator()() const::{lambda(c10::Half)#1}, std::array<char*, 2ul>)
```

After:
```
void at::native::vectorized_elementwise_kernel<8, at::native::(anonymous namespace)::launch_clamp_scalar(at::TensorIteratorBase&, c10::Scalar, c10::Scalar, at::native::detail::ClampLimits)::{lambda()#1}::operator()() const::{lambda()#9}::operator()() const::{lambda(c10::BFloat16)#1}, std::array<char*, 2ul> >(int, at::native::(anonymous namespace)::launch_clamp_scalar(at::TensorIteratorBase&, c10::Scalar, c10::Scalar, at::native::detail::ClampLimits)::{lambda()#1}::operator()() const::{lambda()#9}::operator()() const::{lambda(c10::BFloat16)#1}, std::array<char*, 2ul>)
void at::native::vectorized_elementwise_kernel<8, at::native::(anonymous namespace)::launch_clamp_scalar(at::TensorIteratorBase&, c10::Scalar, c10::Scalar, at::native::detail::ClampLimits)::{lambda()#1}::operator()() const::{lambda()#8}::operator()() const::{lambda(c10::Half)#1}, std::array<char*, 2ul> >(int, at::native::(anonymous namespace)::launch_clamp_scalar(at::TensorIteratorBase&, c10::Scalar, c10::Scalar, at::native::detail::ClampLimits)::{lambda()#1}::operator()() const::{lambda()#8}::operator()() const::{lambda(c10::Half)#1}, std::array<char*, 2ul>)
```

cc @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia @manuelcandales @angelayi

### CODE REVIEW

**Review by eqy — COMMENTED** [review]
For completeness, should we also add 
```
1000 == __CUDA_ARCH_FAMILY_SPECIFIC__
```
in case some users build with e.g., `TORCH_CUDA_ARCH_LIST=10.0f`?

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 1]
```diff
@@ -166,7 +166,7 @@ template <int vec_size, typename func_t, typename array_t>
 C10_LAUNCH_BOUNDS_1(num_threads())
 __global__ void vectorized_elementwise_kernel(int N, func_t f, array_t data) {
   if constexpr (vec_size == 8) {
-#if __CUDA_ARCH__ == 900 || __CUDA_ARCH__ == 1000
+#if __CUDA_ARCH__ == 900 || __CUDA_ARCH__ == 1000 || __CUDA_ARCH__ == 1030
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/CUDALoops.cuh]: can we make it a bit more future proof, so that next time we don't have to painfully compare kernel names?

Comment (Aidyn-A) [reply aten/src/ATen/native/cuda/CUDALoops.cuh]: Sure, I have made it work for all `10.x`. It is not that beneficial to compile `vec8` on `11.0` and `12.x`, so I omitted them. I will remove all conditions on `__CUDA_ARCH__` when we no longer need to maintain CUDA 12.x, so we can take advantage of the binary size compression in CUDA 13+ builds.

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 2]
```diff
@@ -166,7 +166,7 @@ template <int vec_size, typename func_t, typename array_t>
 C10_LAUNCH_BOUNDS_1(num_threads())
 __global__ void vectorized_elementwise_kernel(int N, func_t f, array_t data) {
   if constexpr (vec_size == 8) {
-#if __CUDA_ARCH__ == 900 || __CUDA_ARCH__ == 1000
+#if __CUDA_ARCH__ == 900 || __CUDA_ARCH__ / 100 == 10 || __CUDA_ARCH_FAMILY_SPECIFIC__ == 1000
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/CUDALoops.cuh]: I don't think we need `__CUDA_ARCH_FAMILY_SPECIFIC__` macro here because the kernel doesn't use any family-specific instructions, and regular `__CUDA_ARCH__` will be set even if someone compiles with `1xxf` https://docs.nvidia.com/cuda/cuda-programming-guide/05-appendices/cpp-language-extensions.html#cuda-arch-specific-and-cuda-arch-family-specific

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 3]
```diff
@@ -308,8 +308,7 @@ static inline void launch_vectorized_kernel(
   // that causes some numerical mismatches with uint8 on sm80 and sm90.
   // TODO: Revisit this after CUDA 12.8 update.
   cudaDeviceProp* p = at::cuda::getDeviceProperties(stream.device().index());
-  const int computeCapability = p->major * 10 + p->minor;
-  if (computeCapability != 90 && computeCapability != 100) {
+  if (p->major != 9 && p->major != 10) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/CUDALoops.cuh:313]: If this line and ifdef in the kernel go out of sync, and you keep `vec_size` 8 for an arch that's excluded by ifdef, you'll get an empty kernel. Can you find a way to make sure this doesn't happen? Even

Comment (Aidyn-A) [reply aten/src/ATen/native/cuda/CUDALoops.cuh:313]: There are two things I can do simultaneously:

1. Raise an error from the kernel if it attempts to call an empty part of kernel.
2. Implement a custom linter that compares this line vs `#if __CUDA_ARCH__ ...` line above, so the numbers must match.

Or just add a big fat comment, so anyone who modifies this file must keep the arches in sync.

Comment (ngimel) [reply aten/src/ATen/native/cuda/CUDALoops.cuh:313]: Adding device assert in the kernel should suffice

**Review by ngimel — APPROVED** [review]
